python/mpt/src/mpt/state_diff.py

```diff
@@ -0,0 +1,243 @@
+import logging
+from typing import Dict, Optional
+
+from ethereum.cancun.fork_types import EMPTY_ACCOUNT, Account, Address
+from ethereum.cancun.state import State, get_account, get_storage
+from ethereum_types.bytes import Bytes32
+from ethereum_types.numeric import U256
+
+# Set up logger
+logger = logging.getLogger("mpt.state_diff")
+
+
+class AccountDiff:
+    """
+    A class that contains the differences for a single account, including its storage changes.
+    If account is None, it represents a deletion.
+    """
+
+    account: Optional[Account]
+    storage_updates: Dict[Bytes32, U256]
+
+    def __init__(self, account: Optional[Account] = None):
+        self.account: Optional[Account] = account
+        self.storage_updates: Dict[Bytes32, U256] = {}
+
+
+class StateDiff:
+    """
+    A class that contains the differences between two states.
+    """
+
+    account_diffs: Dict[Address, AccountDiff]
+
+    def __init__(self):
+        self.account_diffs: Dict[Address, AccountDiff] = {}
+
+    @classmethod
+    def from_pre_post(cls, pre_state: State, post_state: State) -> "StateDiff":
+        """
+        Compute a state diff between two states: a pre_state and a post_state.
+
+        Parameters
+        ----------
+        pre_state : State
+            The state before executing a block
+        post_state : State
+            The state after executing a block
+
+        Returns
+        -------
+        StateDiff
+            A StateDiff object containing:
+            - 'account_diffs': Dict[Address, AccountDiff] - Accounts that were added, modified or deleted with their storage changes
+        """
+        return compute_state_diff(pre_state, post_state)
+
+
+def compute_state_diff(pre_state: State, post_state: State) -> StateDiff:
+    """
+    Compute the difference between pre_state and post_state.
+
+    Parameters
+    ----------
+    pre_state : State
+        The state before executing a block
+    post_state : State
+        The state after executing a block
+
+    Returns
+    -------
+    StateDiff
+        A StateDiff object containing:
+        - 'account_diffs': Dict[Address, AccountDiff] - Accounts that were added, modified or deleted with their storage changes
+    """
+
+    diff = StateDiff()
+
+    pre_addresses = set(pre_state._main_trie._data.keys())
+    post_addresses = set(post_state._main_trie._data.keys())
+
+    logger.debug(
+        f"Computing state diff: {len(pre_addresses)} pre addresses, {len(post_addresses)} post addresses"
+    )
+
+    # Find deleted accounts
+    deletions = pre_addresses - post_addresses
+    if deletions:
+        logger.error(
+            f"Warning: Since EIP-6780, we should not see any accounts marked for deletion at the level of the block - found {len(deletions)}"
```

Comment:
should this exit rather than log ?

Reply:
i've noticed edge cases where an empty account is in the pre-state and it is deleted in the post state

i could catch that case and exit if it's not in this case

Reply:
I don't think that any such check belongs to a state diff function. this is rather a post processing check, that may or may not be used depending on the state's block